fix: nonfatal missing key commands (resolves #6397)j

ariane-emory opened the pull request:
This PR changes the behaviour so that binding missing key commands in a user's config is not a fatal error that prevents the startup of opencode entirely.

With this PR, keybindings for non-existent keywords, when struck will instead display a toast saying "No command 'the_fake_command_name'".

Resolves #6397.

rekram1-node commented on packages/opencode/src/provider/provider.ts:
```diff
@@ -48,15 +48,15 @@ export namespace Provider {
+    "@ai-sdk/cohere": createCohere as any,
+    "@ai-sdk/gateway": createGateway as any,
+    "@ai-sdk/togetherai": createTogetherAI as any,
+    "@ai-sdk/perplexity": createPerplexity as any,
```
why did u make this change?

ariane-emory replied:
Yikes, accidental. Will fix presently!

ariane-emory replied:
@rekram1-node Reverted this accidental change.

rekram1-node commented:
this shouldn't be enforced on the tui side, I would expect the tui to always be fed a valid config, and maybe u can introduce a config event for warnings that you emit that the tui could display

ariane-emory commented:
@rekram1-node Alright, I'll look into how that could work and see what I can come up with.

ariane-emory committed:
fix: make unknown keybind commands non-fatal with warnings

- Change Keybinds schema from .strict() to .passthrough() to allow unknown keys
- Add ValidKeybindNames set for validating keybind names
- Add Config.Warning schema and Config.warnings() function
- Add GET /config/warnings endpoint to fetch warnings
- Add warning toast display in TUI when sync completes
- Add tests for warning generation with unknown keybind names

Fixes #6397

ariane-emory committed:
fix: add toast when pressing keybind for unknown command

When user presses a key that matches a keybind configured for an unknown
command name, show a warning toast with the command name.

ariane-emory committed:
fix: use server-provided unknown keybinds list for toast messages

- Fetch unknown keybinds from /config/warnings endpoint instead of
  iterating all keybinds in config
- Add keybinds array field to Warning schema to provide list of
  unknown keybind names
- Fix TypeScript error from passthrough schema by casting value
- Regenerate SDK with ConfigWarning.keybinds field

ariane-emory committed:
fix: use consistent 'Unknown keybind command' message

ariane-emory committed:
fix: use proper pluralization for unknown keybind commands message

ariane-emory committed:
chore: restore unrelated duplicate block to keep branch tightly scoped

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Fix: Implement nonfatal missing keybind commands with pre-processing approach

- Pre-process unknown keybinds before Zod validation (keep .strict() schema)
- Store unknown keybinds with name+binding in warnings
- Update TUI to parse and match unknown keybinds directly
- Remove unused BusEvent for warnings
- Update tests to verify new warning structure
- Regenerate TypeScript SDK (KeybindsConfig without index signature)

Resolves #6397

ariane-emory committed:
revert: Remove unnecessary null check in keybind parsing

The null check was added during earlier iterations when .passthrough() was used
in the keybind schema. With the current pre-processing approach that strips
unknown keybinds and maintains .strict(), all keybind values have defaults
applied and will never be undefined, making this check unnecessary.

ariane-emory committed:
refactor: Simplify Config.global() access to single line

ariane-emory committed:
chore: remove unused Event.config.warning from openapi schema

ariane-emory committed:
refactor: use BusEvent for config warnings instead of API endpoint

- Add Config.Event.Warning BusEvent definition
- Server sends warning events to clients on SSE connect
- TUI subscribes to config.warning events instead of polling API
- Remove /config/warnings API endpoint (no longer needed)
- Regenerate SDK

This fulfills the maintainer's request to use event-based
communication for config warnings rather than TUI-side enforcement.

ariane-emory commented:
@rekram1-node Well.... it turned out to be a lot trickier to do than I thought it would, but after a few rerolls this does _seem_ to be doing it in the way that you'd requested?

- strips the 'unknown' key commands out prior to validation, so the config always receives a valid config.
- -emits a `config.warning` BusEvent that TUI then consumes and reacts to by displaying a warning like `Unknown key commands: foo, bar`.
- retains the `Unknown key command: foo` toast upon striking one of these keys.

Whaddya think? If there are further refinements you'd like. don't hesitate to ask!

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
revert: remove unnecessary formatting changes from openapi.json

ariane-emory committed:
fix: add ConfigWarning schema to openapi.json for SDK compatibility

ariane-emory committed:
Merge dev into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge dev into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Remove duplicate code in config.ts

ariane-emory committed:
Remove duplicate config loading code

ariane-emory committed:
Merge dev into fix/nonfatal-missing-key-commands

ariane-emory commented:
Requires further testing/revision after significant recent changes in dev, re-drafting until that's been done.

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge dev into fix/nonfatal-missing-key-commands

Resolved conflict in dialog-command.tsx by preserving the unknown keybind
handling logic from fix/nonfatal-missing-key-commands while incorporating
the entries memoization changes from dev.

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands

ariane-emory committed:
Merge dev into fix/nonfatal-missing-key-commands

Resolved conflicts:
- Combined Flag.OPENCODE_DISABLE_PROJECT_CONFIG check from dev with unknownKeybinds collection from current branch in config.ts

ariane-emory committed:
Fix TypeScript namespace conflict between imported ServerEvent and local Config.Event

ariane-emory committed:
Merge branch 'dev' into fix/nonfatal-missing-key-commands
